launcher/src/main/java/org/apache/spark/launcher/LauncherServer.java

```diff
@@ -376,9 +353,23 @@ public void close() throws IOException {
         clients.remove(this);
       }
 
-      synchronized (this) {
-        super.close();
-        notifyAll();
+      super.close();
+    }
+
+    /**
+     * Close the connection and wait for any buffered data to be processed before returning.
+     * This ensures any changes reported by the child application take effect.
+     */
+    public void closeAndWait() throws IOException {
+      close();
+
+      Thread connThread = this.connectionThread;
+      if (Thread.currentThread() != connThread) {
```

Comment:
Is this a safeguard or it can really happen? i.e. the connection thread calls `closeAndWait`.

Reply:
More of a safeguard. I don't think it would happen in this version of the code, but better be safe.